atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/MarkAndCasConsensusForgettingStore.java

```diff
@@ -112,12 +133,143 @@ public void put(Map<Cell, byte[]> values) {
         kvs.setOnce(tableRef, values);
     }
 
-    private byte[] getRowName(Map<Cell, byte[]> values) {
-        Set<ByteBuffer> rows = values.keySet().stream()
-                .map(Cell::getRowName)
-                .map(ByteBuffer::wrap)
-                .collect(Collectors.toSet());
-        Preconditions.checkState(rows.size() == 1, "Only allowed to make batch updates to cells across one row.");
-        return Iterables.getOnlyElement(rows).array();
+    @SuppressWarnings("ByteBufferBackingArray")
+    @VisibleForTesting
+    static void processBatch(KeyValueService kvs, TableReference tableRef, List<BatchElement<CasRequest, Void>> batch) {
+        List<BatchElement<CasRequest, Void>> pendingUpdates = filterOptimalUpdatePerCell(batch);
+
+        Map<ByteBuffer, List<BatchElement<CasRequest, Void>>> pendingRawRequests = pendingUpdates.stream()
+                .collect(Collectors.groupingBy(
+                        elem -> ByteBuffer.wrap(elem.argument().cell().getRowName())));
+
+        // There is one request per row
+        Map<CasRequest, CasResponse> resultMap = new HashMap<>();
+
+        for (Map.Entry<ByteBuffer, List<BatchElement<CasRequest, Void>>> requestEntry : pendingRawRequests.entrySet()) {
+            ByteBuffer rowName = requestEntry.getKey();
+            List<BatchElement<CasRequest, Void>> pendingRequests = requestEntry.getValue();
+            MultiCheckAndSetRequest multiCheckAndSetRequest =
+                    multiCASRequest(tableRef, rowName.array(), pendingRequests);
+            try {
+                kvs.multiCheckAndSet(multiCheckAndSetRequest);
+                pendingRequests.forEach(req -> resultMap.put(req.argument(), CasResponse.success()));
+            } catch (Exception e) {
+                pendingRequests.forEach(req -> resultMap.put(req.argument(), CasResponse.failure(e)));
+            }
+        }
+
+        populateResult(batch, resultMap);
+    }
+
+    private static void populateResult(
+            List<BatchElement<CasRequest, Void>> batch, Map<CasRequest, CasResponse> results) {
+        CasResponse defaultResponse =
+                CasResponse.failure(new CheckAndSetException("There were one or more concurrent updates for the same "
+                        + "cell and a higher ranking update was selected to be executed. "));
+        batch.forEach(elem -> {
+            CasResponse response = results.getOrDefault(elem.argument(), defaultResponse);
+            if (response.successful()) {
+                elem.result().set(null);
+            } else {
+                elem.result().setException(response.exception().get());
```

Comment:
One possible issue here is that we don't have detail on failure. It seems excessively difficult to have some sensible detail here so it's probably fine. @jeremyk-91 ?

Also, the exception type does not match the declared exception -- we need to fix that